src/components/ingestion/AbstractIngestionEngine.js

```diff
@@ -1,100 +1,209 @@
+	 * @param  {String}   enc     The encoding (part of the Node Stream API)
+	 * @param  {Function} done    The callback for when the processing is complete
+	 */
+	ingestPayload = (payload, enc, done) => {
+		if (!(payload instanceof Payload)) {
+			done(new Error('ingestPayloadStream received non-Payload data'))
+		} else {
+			done(this.producer.send({
+				topic: dataTypes.STREAM.CONTAINER,
+				messages: [{
+					value: Payload.serialize(payload),
+				}],
+			}))
+		}
 	}
 
 	/**
-	 * Ingests data into the TV Kitchen pipeline.
+	 * Updates ingestion state based on the latest demuxed packet data.
+	 *
+	 * This method is called by our MPEG-TS demuxer, and allows the ingestion engine to track
+	 * the most recent demuxed packet.
 	 *
-	 * @param  {Buffer} data The buffer of data to be ingested.
-	 * @return {Boolean}     Whether or not the buffer was successfully ingested.
+	 * @param  {Packet} packet The latest TSDemuxer Packet object.
 	 */
-	ingestData = async (data) => {
-		const payload = new Payload()
-		payload.data = data
-		await this.producer.send({
-			topic: dataTypes.STREAM.DATA,
-			messages: [{
-				value: data,
-			}],
-		})
-		return true
+	onDemuxedPacket = (packet) => {
+		this.mostRecentDemuxedPacket = packet
 	}
 
 	/**
-	 * Continuously ingest of a stream of data.
+	 * Returns the most recent coherent stream packet processed by the ingestion engine.
+	 * This packet is lower level than the MPEG-TS container, and represents an audio or video
+	 * packet demuxed from the MPEG-TS stream.
+	 *
+	 * @return {Packet} The most recent Packet object extracted by TSDemuxer.
+	 */
+	getMostRecentDemuxedPacket = () => this.mostRecentDemuxedPacket
+
+	/**
+	 * Process a new chunk of data from an MPEG-TS stream. The chunks passed to this
+	 * function should be presented sequentially, and combine to form a coherent MPEG-TS
+	 * data stream, but they can be of arbitrary size.
+	 *
+	 * This method is called by a NodeJS Transform stream and so matches that spec
+	 *
+	 * @param  {Buffer} mpegtsData The latest sequential chunk of MPEG-TS data.
+	 * @param  {String} enc        The encoding of the passed data.
+	 * @param  {Function(err,result)} done     A `(err, result) => ...` callback.
 	 *
-	 * @param  {ReadableStream} readableStream The stream of data to be recorded.
-	 * @return {ReadableStream}                The input stream (for chaining)
 	 */
-	recordStream = (readableStream) => readableStream
-		.on('data', async (data) => {
-			readableStream.pause()
-			await this.ingestData(data)
-			readableStream.resume()
+	processMpegtsStreamData = (mpegtsData, enc, done) => {
+		this.mpegtsDemuxer.process(mpegtsData)
+		const demuxedPacket = this.getMostRecentDemuxedPacket() || generateEmptyPacket()
+		const payload = new Payload({
+			data: mpegtsData,
+			type: dataTypes.STREAM.CONTAINER,
+			duration: 0,
+			position: tsToMilliseconds(demuxedPacket.pts),
+			createdAt: Date.now(),
```

Comment:
This emits a UNIX epoch time, which doesn't match our current [Payload definition](https://github.com/tvkitchen/base/blob/master/packages/classes/docs/PAYLOAD.md#attributes), which says it should be ISO 8601. (Smart folks, these `Payload` craftspeople.)

Reply:
Good call.  As an aside, I started exploring how to bake type checks into Payload for that kind of thing (this will NOT be the first time we try to assign the wrong thing to createdAt.  In particular I imagine that we'll accidentally assign date objects instead of ISO strings...)